### DISCUSSION TITLE
How to get the syntax tree from an open document?

### DISCUSSION THREAD

**Discussion by CPTANT** [microsoft/vscode-discussions#525] [Extension Development QnA] [answered]
Hi guys, 

I am working on a visual studio code extension and for my extension I need to access some properties of all symbol definitions in an open document. I am currently doing this by using regex, but this is actually an unsuited tool for the job, a proper parser would be far more ideal. I know that many of the code analyzers must be using something like this to create their warnings etc. Is there any way to use the same system they are using or to create a syntax tree yourself?

I start with an open document of code and try to get a syntax tree for use within typescript for visual studio code extension development.

**Comment by alefragnani** [comment, marked as answer]
HI @CPTANT ,

Unfortunately, there is no such thing. VS Code API doesn’t provide the syntax tree of the document. It depends entirely on the language extension you have opened.

You can however, retrieve _all the symbols_ within the opened document. This will provides to you the same data you see in the `Outline` view. If this fit your needs, you could use the `executeDocumentSymbolProvider ` command, like this:

```ts
    const docSymbols = await commands.executeCommand(
        'vscode.executeDocumentSymbolProvider',
        window.activeTextEditor.document.uri
    ) as DocumentSymbol[];
```

> extracted from https://github.com/alefragnani/vscode-separators/blob/629d366c9ec02a252857271ff33e7147aaa9809b/src/symbols.ts#L56

Hope this helps
